subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -121,41 +119,43 @@ public SnapshotHierarchy updateVfsOnBuildFinished(SnapshotHierarchy root, WatchM
     }
 
     @Override
-    public Collection<Path> getWatchedHierarchies() {
+    public Collection<File> getWatchedRoots() {
         return watchedHierarchies.getWatchedRoots();
     }
 
     private void updateWatchedHierarchies(SnapshotHierarchy root) {
-        Set<Path> oldWatchedRoots = watchedHierarchies.getWatchedRoots();
-        watchedHierarchies.updateWatchedHierarchies(watchableHierarchies, root);
-        Set<Path> newWatchedRoots = watchedHierarchies.getWatchedRoots();
+        Set<File> oldWatchedRoots = watchedHierarchies.getWatchedRoots();
+        watchedHierarchies = WatchedHierarchies.resolveWatchedHierarchies(watchableHierarchies, root);
+        Set<File> newWatchedRoots = watchedHierarchies.getWatchedRoots();
 
         if (newWatchedRoots.isEmpty()) {
             LOGGER.info("Not watching anything anymore");
         }
-        Set<Path> hierarchiesToStopWatching = new HashSet<>(oldWatchedRoots);
-        Set<Path> hierarchiesToStartWatching = new HashSet<>(newWatchedRoots);
+        Set<File> hierarchiesToStopWatching = new HashSet<>(oldWatchedRoots);
+        Set<File> hierarchiesToStartWatching = new HashSet<>(newWatchedRoots);
         hierarchiesToStopWatching.removeAll(newWatchedRoots);
         hierarchiesToStartWatching.removeAll(oldWatchedRoots);
         if (hierarchiesToStartWatching.isEmpty() && hierarchiesToStopWatching.isEmpty()) {
             return;
         }
         if (!hierarchiesToStopWatching.isEmpty()) {
-            fileWatcher.stopWatching(hierarchiesToStopWatching.stream()
-                .map(Path::toFile)
-                .collect(Collectors.toList())
-            );
+            fileWatcher.stopWatching(hierarchiesToStopWatching);
         }
         if (!hierarchiesToStartWatching.isEmpty()) {
             fileWatcher.startWatching(hierarchiesToStartWatching.stream()
-                .map(Path::toFile)
                 .peek(locationToWatchValidator::validateLocationToWatch)
+                .peek(watchableHierarchies::armWatchProbe)
```

Comment:
IIUC, we now only arm the watch probe for the hierarchical watcher and not for the non-hierarchical watchers? So that is something still missing from the PR?

For the non-hierarchical watcher, we also would need to make sure to actually start watching the parent directory of the probe file, right?

Reply:
Yes, that's still missing. I added it to the todo in the issue description.